src/dataflow/decode/mod.rs

```diff
@@ -133,14 +135,17 @@ where
             stream,
             &enc.value_schema,
             enc.schema_registry_url,
-            envelope == Envelope::Debezium,
+            envelope.get_avro_envelope_type() == interchange::avro::EnvelopeType::Debezium,
         ),
         (DataEncoding::AvroOcf { .. }, _) => {
             unreachable!("Internal error: Cannot decode Avro OCF separately from reading")
         }
         (_, Envelope::Debezium) => unreachable!(
             "Internal error: A non-Avro Debezium-envelope source should not have been created."
         ),
+        (_, Envelope::Upsert(_)) => unreachable!(
+            "Internal error: A non-Kafka Upsert-envelope source should not have been created."
```

Comment:
Here you seem to be checking the format (Avro), not the connector (Kafka), but your error message talks about Kafka.

Reply:
This is a case of incomplete PR splitting. In #2493, kafka streams (the only place where upsert is supported) go to a separate method from this because kafka streams become `Stream<G, ((Row, Option<Row>), Timestamp, Diff)>` instead of `Stream<G, (Row, Timestamp, Diff)>`. But you do raise a good point, and in #2493, I will move the `(_, Envelope::Upsert(_))` case to be the first case in this method so that even `(Avro, Envelope::Upsert(_)` will be unreachable. And I'll also change the error message here to "Upsert is not yet implemented."